File: app/src/main/java/com/whiteside/cafe/WishList/WishListPresenter.java
package com.whiteside.cafe.WishList;

import androidx.annotation.NonNull;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.Timestamp;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QuerySnapshot;

import java.util.Map;

import Model.Item;
import Model.Product;

public class WishListPresenter {

    private final FirebaseFirestore fStore;
    private final FirebaseAuth auth;
    OnGetWishListItem onGetWishListItem;
    OnAddWishListItem onAddWishListItem;
    OnRemoveWishListItemListener onRemoveWishListItemListener;


    public WishListPresenter() {
        fStore = FirebaseFirestore.getInstance();
        auth = FirebaseAuth.getInstance();
    }


    public void setOnGetWishListItem(OnGetWishListItem onGetWishListItem) {
        this.onGetWishListItem = onGetWishListItem;
    }

    public void setOnAddWishListItem(OnAddWishListItem onAddWishListItem) {
        this.onAddWishListItem = onAddWishListItem;
    }

    public void setOnRemoveWishListItemListener(OnRemoveWishListItemListener onRemoveWishListItemListener) {
        this.onRemoveWishListItemListener = onRemoveWishListItemListener;
    }


    private void getItemsFromFirebase() {
        FirebaseFirestore fStore = FirebaseFirestore.getInstance();
        FirebaseAuth auth = FirebaseAuth.getInstance();

        fStore.collection("Users")
                .document(auth.getUid())
                .collection("WishList")
                .get()
                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<QuerySnapshot> task) {
                        if (task.isSuccessful()) {
                            if (task.getResult().isEmpty()) {
                                onGetWishListItem.onWishListIsEmpty();
                            }

                            for (DocumentSnapshot ds : task.getResult().getDocuments()) {
                                onGetWishListItem.onGetWishListItemSuccess(ds.toObject(Item.class));
                            }
                        } else {
                            onGetWishListItem.onGetWishListItemFailed(task.getException());
                        }
                    }
                });
    }

    public void getWishListItems() {
        getItemsFromFirebase();
    }

    public void addItem(Product product) {
        Map<String, Integer> wishList = product.getWishes();
        wishList.put(auth.getUid(), 1);
        product.setWishes(wishList);

        Thread thread2 = new Thread() {
            @Override
            public void run() {
                fStore.collection("Categories")
                        .document(product.getCategoryName())
                        .collection("Products")
                        .document(product.getProductId())
                        .update("wishes", wishList)
                        .addOnSuccessListener(new OnSuccessListener<Void>() {
                            @Override
                            public void onSuccess(Void aVoid) {
                                onAddWishListItem.onAddWishListItemSuccess();
                            }
                        });
            }
        };


        Item item = new Item();
        item.setTime(Timestamp.now());
        item.setCategoryName(product.getCategoryName());
        item.setProductId(product.getProductId());
        item.setQuantity(1);

        Thread thread1 = new Thread() {
            @Override
            public void run() {
                fStore.collection("Users")
                        .document(auth.getUid())
                        .collection("WishList")
                        .document(product.getCategoryName() + product.getProductId())
                        .set(item)
                        .addOnSuccessListener(new OnSuccessListener<Void>() {
                            @Override
                            public void onSuccess(Void aVoid) {
                                thread2.start();
                            }
                        });
            }
        };

        thread1.start();
    }

    public void removeItem(Product product) {
        Map<String, Integer> wishList = product.getWishes();
        wishList.remove(auth.getUid());

        product.setWishes(wishList);

        Thread thread2 = new Thread() {
            @Override
            public void run() {
                fStore.collection("Categories")
                        .document(product.getCategoryName())
                        .collection("Products")
                        .document(product.getProductId())
                        .update("wishes", wishList)
                        .addOnSuccessListener(new OnSuccessListener<Void>() {
                            @Override
                            public void onSuccess(Void aVoid) {
                                onRemoveWishListItemListener.onRemoveWishListItemSuccess();
                            }
                        });
            }
        };

        Thread thread1 = new Thread() {
            @Override
            public void run() {
                fStore.collection("Users")
                        .document(auth.getUid())
                        .collection("WishList")
                        .document(product.getCategoryName() + product.getProductId())
                        .delete()
                        .addOnSuccessListener(new OnSuccessListener<Void>() {
                            @Override
                            public void onSuccess(Void aVoid) {
                                thread2.start();
                            }
                        });
            }
        };

        thread1.start();
    }
}

File: app/src/main/java/com/whiteside/cafe/ui/wishList/WishListPresenter.kt
package com.whiteside.cafe.ui.wishList

import com.google.firebase.Timestamp
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.whiteside.cafe.model.Item
import com.whiteside.cafe.model.Product

class WishListPresenter {
    private val fStore: FirebaseFirestore?
    private val auth: FirebaseAuth?
    var onGetWishListItem: OnGetWishListItem? = null
    var onAddWishListItem: OnAddWishListItem? = null
    var onRemoveWishListItemListener: OnRemoveWishListItemListener? = null
    fun setOnGetWishListItem(onGetWishListItem: OnGetWishListItem?) {
        this.onGetWishListItem = onGetWishListItem
    }

    fun setOnAddWishListItem(onAddWishListItem: OnAddWishListItem?) {
        this.onAddWishListItem = onAddWishListItem
    }

    fun setOnRemoveWishListItemListener(onRemoveWishListItemListener: OnRemoveWishListItemListener?) {
        this.onRemoveWishListItemListener = onRemoveWishListItemListener
    }

    private fun getItemsFromFirebase() {
        val fStore = FirebaseFirestore.getInstance()
        val auth = FirebaseAuth.getInstance()
        fStore.collection("Users")
            .document(auth.uid)
            .collection("WishList")
            .get()
            .addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    if (task.result.isEmpty) {
                        onGetWishListItem.onWishListIsEmpty()
                    }
                    for (ds in task.result.documents) {
                        onGetWishListItem.onGetWishListItemSuccess(ds.toObject(Item::class.java))
                    }
                } else {
                    onGetWishListItem.onGetWishListItemFailed(task.exception)
                }
            }
    }

    fun getWishListItems() {
        getItemsFromFirebase()
    }

    fun addItem(product: Product?) {
        val wishList = product.getWishes()
        wishList[auth.getUid()] = 1
        product.setWishes(wishList)
        val thread2: Thread = object : Thread() {
            override fun run() {
                fStore.collection("Categories")
                    .document(product.getCategoryName())
                    .collection("Products")
                    .document(product.getProductId())
                    .update("wishes", wishList)
                    .addOnSuccessListener { onAddWishListItem.onAddWishListItemSuccess() }
            }
        }
        val item = Item()
        item.time = Timestamp.now()
        item.categoryName = product.getCategoryName()
        item.productId = product.getProductId()
        item.quantity = 1
        val thread1: Thread = object : Thread() {
            override fun run() {
                fStore.collection("Users")
                    .document(auth.getUid())
                    .collection("WishList")
                    .document(product.getCategoryName() + product.getProductId())
                    .set(item)
                    .addOnSuccessListener { thread2.start() }
            }
        }
        thread1.start()
    }

    fun removeItem(product: Product?) {
        val wishList = product.getWishes()
        wishList.remove(auth.getUid())
        product.setWishes(wishList)
        val thread2: Thread = object : Thread() {
            override fun run() {
                fStore.collection("Categories")
                    .document(product.getCategoryName())
                    .collection("Products")
                    .document(product.getProductId())
                    .update("wishes", wishList)
                    .addOnSuccessListener { onRemoveWishListItemListener.onRemoveWishListItemSuccess() }
            }
        }
        val thread1: Thread = object : Thread() {
            override fun run() {
                fStore.collection("Users")
                    .document(auth.getUid())
                    .collection("WishList")
                    .document(product.getCategoryName() + product.getProductId())
                    .delete()
                    .addOnSuccessListener { thread2.start() }
            }
        }
        thread1.start()
    }

    init {
        fStore = FirebaseFirestore.getInstance()
        auth = FirebaseAuth.getInstance()
    }
}